numpy/core/src/multiarray/alloc.c

```diff
@@ -222,10 +222,8 @@ PyDataMem_NEW(size_t size)
     if (_PyDataMem_eventhook != NULL) {
         NPY_ALLOW_C_API_DEF
         NPY_ALLOW_C_API
-        if (_PyDataMem_eventhook != NULL) {
-            (*_PyDataMem_eventhook)(NULL, result, size,
-                                    _PyDataMem_eventhook_user_data);
-        }
+        (*_PyDataMem_eventhook)(NULL, result, size,
+                                _PyDataMem_eventhook_user_data);
```

Comment:
These changes aren't safe. This is https://en.wikipedia.org/wiki/Double-checked_locking, because `NPY_ALLOW_C_API` is essentially `PyGILState_Ensure()`

Reply:
Cool, never heard of it -- the linked article isn't exactly a big endorsement though:

> The pattern, when implemented in some language/hardware combinations, can be unsafe. At times, it can be considered an anti-pattern

Reply:
All I can find arguing against it is that it's not suitable for safe static initialization. We're using it in the following context though:
```C
// only acquire the GIL if we need ti
if (_python_callback) {
    acquire_gil();
    // check that the callback was not deregistered while waiting for the GIL
    if (_python_callback) {
        _python_callback()
    }
    release_gil();
}
```
If we wanted to drop the idiom, it would become the slower
```C
// only acquire the GIL if we need ti
acquire_gil();
// check that the callback was not deregistered while waiting for the GIL
if (_python_callback) {
    _python_callback()
}
release_gil();
```

Removing the outer `if` is a performance hit, removing the inner `if` is completely unsafe.